examples/settings/settings_main.c

```diff
@@ -0,0 +1,276 @@
+#  endif
+  if (stat(CONFIG_LIBC_TMPDIR, &sbuf))
+    {
+      ret = nx_mount(NULL, CONFIG_LIBC_TMPDIR, "tmpfs", 0, NULL);
+      if (ret < 0)
+        {
+          printf("ERROR: Failed to mount tmpfs at %s: %d\n",
+                  CONFIG_LIBC_TMPDIR, ret);
+          goto end;
+        }
+    }
+
+  strcpy(path, CONFIG_LIBC_TMPDIR);
+  strcat(path, "/settings");
+#else
+  strcpy(path, CONFIG_EXAMPLES_SETTINGS_EXISTING_STORAGE);
+  if (path == NULL)
+    {
+      printf("Settings filepath is empty!");
+      goto end;
+    }
+
+  strcat(path, "/settings");
+#endif
+
+  printf("Example of settings usage: %s. Path: %s\n",
+         ((storage_type == STORAGE_TEXT) ? "text" : "binary"), path);
+  printf("-------------------------------------------------------\n");
+
+  ret = settings_init();
+  if (ret < 0)
+    {
+      printf("settings init failed: %d", ret);
+      goto end;
+    }
+
+  ret = settings_setstorage(path, storage_type);
+  if (ret == -ENOENT)
+    {
+      printf("INFO: no existing binary settings file found. Creating it.\n");
+      fd = open(path, O_CREAT);
+      if (fd < 0)
+        {
+          printf("failed to create settings file\n");
+          goto end;
+        }
+
+      close(fd);
+    }
+  else if (ret < 0)
+    {
+      printf("settings setstorage failed: %d\n", ret);
+      goto end;
+    }
+  else
+    {
+      printf("INFO: a settings file was found and loaded\n");
+      ret = settings_get("v1", SETTING_STRING, readstr,
+                         CONFIG_SETTINGS_VALUE_SIZE);
+      if (ret < 0)
+        {
+          printf("settings retrieve failed: %d\n", ret);
+          goto end;
+        }
+
+      printf("Retrieved settings value (v1) with value:%s\n",
+              readstr);
+    }
+
+  printf("Creating an integer settings value (v1) with value:0x%x\n",
+          testval);
+  ret = settings_create("v1", SETTING_INT, testval);
+  if (ret < 0)
+    {
+      printf("settings create failed: %d\n", ret);
+      goto end;
+    }
+
+  ret = settings_type("v1", &stype);
+  if (ret < 0)
+    {
+      printf("failed to read settings type: %d\n", ret);
+      goto end;
+    }
+
+  printf("Retrieved setting type is:%d\n", stype);
+
+  ret = settings_get("v1", SETTING_INT, &testval);
+  if (ret < 0)
+    {
+      printf("settings retrieve failed: %d\n", ret);
+      goto end;
+    }
+
+  printf("Retrieved integer settings value (v1) with value:0x%x\n",
+          testval);
+
+  ret = settings_get("v1", SETTING_BOOL, &testval);
+  if (ret < 0)
+    {
+      printf("Deliberate fail: BOOL type requested not INT. Error:%d\n",
+              ret);
+    }
+
+  ret = settings_get("v2", SETTING_INT, &testval);
+  if (ret < 0)
+    {
+      printf("Deliberate fail: non-existent setting requested. Error:%d\n",
+             ret);
+    }
+
+  teststr = "I'm a string!";
+  printf("Changing an integer settings value (v1) to string:%s\n",
+          teststr);
+  ret = settings_create("v1", SETTING_STRING, teststr);
```

Comment:
This should always fail. You are trying to create a setting that already exists!

If you want to change its type, you should be able to just `settings_set()` it to the new type.

Reply:
I was rather left to guess what the functions did, and reverse engineer it all, so I did my best as a first pass :)

To me "create" is the first time the app wants to make use of a setting and "set" is to set the value and "get" is to get the value. If you try and "set" or "get" a value of the wrong type, should it not fail? But, yes, if you do want to change the type what do we do? I chose "create" meaning, in this case, re-create.

It does work, as I have interpreted it, but if its not in line with what you intended I apologise, and please clarify/explain :)